### PR TITLE
Add player elytra boost API

### PR DESCRIPTION
**Pull request by Trigary**
#3138

### CODE REVIEW

File: Spigot-API-Patches/0196-Player-elytra-boost-API.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -1635,6 +1635,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+      */
+     public void openBook(@NotNull ItemStack book);
+ 
++    // Paper start
++    /**
++     * Boost a Player that's {@link #isGliding()} using a {@link Firework}.
++     * The creation of the entity can get cancelled, no boosting is done in that case.
```
Comment (Proximyst) [inline Spigot-API-Patches/0196-Player-elytra-boost-API.patch]: `If the creation of the entity is cancelled, no boosting is done.` would perhaps sound more natural? I can at least see the current wording confusing non-native english speakers.

Comment (Trigary) [reply Spigot-API-Patches/0196-Player-elytra-boost-API.patch]: Sounds good! I shall make that change sometime in the future.

File: Spigot-API-Patches/0196-Player-elytra-boost-API.patch [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
++     * This method does not fire {@link com.destroystokyo.paper.event.player.PlayerElytraBoostEvent}.
++     *
++     * @param firework The {@link Material#FIREWORK_ROCKET} to boost the player with
++     * @return The {@link Firework} boosting the Player or null if the spawning of the entity was cancelled
++     * @throws IllegalArgumentException if {@link #isGliding()} is false
++     * or if the {@code firework} isn't a {@link Material#FIREWORK_ROCKET}
++     */
++    @Nullable
++    Firework boostElytra(@NotNull ItemStack firework);
++    // Paper end
```
Comment (Machine-Maker) [inline Spigot-API-Patches/0196-Player-elytra-boost-API.patch]: You probably don't need to end the Paper comment section and just remove the following paper start.

File: Spigot-Server-Patches/0477-Player-elytra-boost-API.patch [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
++        Validate.isTrue(firework.getType() == Material.FIREWORK_ROCKET, "Firework must be Material.FIREWORK_ROCKET");
++        
++        net.minecraft.server.ItemStack item = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(firework);
++        net.minecraft.server.World world = ((CraftWorld) getWorld()).getHandle();
++        net.minecraft.server.EntityFireworks entity = new net.minecraft.server.EntityFireworks(world, item, getHandle());
++        return world.addEntity(entity)
++                ? (org.bukkit.entity.Firework) entity.getBukkitEntity()
++                : null;
++    }
++    // Paper end
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0477-Player-elytra-boost-API.patch]: same as above

**Comment by TheFruxz** [comment]
I find this thing incredibly good!

**Review by electronicboy — DISMISSED** [review]

File: Spigot-API-Patches/0196-Player-elytra-boost-API.patch (hunk 2 above)
Comment (MiniDigger) [reply Spigot-API-Patches/0196-Player-elytra-boost-API.patch]: moved the method to the bottom of the paper additions

**Comment by MiniDigger** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
sorry, a bit late, but thanks for your contribution. I moved your additions to the end of the paper addition in the two classes and rebased it

**Pull request merged**